### PR TITLE
Development cycle metric - Part 1 of 2

### PR DESCRIPTION
**Pull request by enzofab91**
## What does this PR do?
This PR is one part of the full feature and mainly is to obtain the data needed to calculate development cycle metric. Next steps includes the calculation of the metric and adding this information in the frontend.

This PR fetch new information of Jira issues when it runs `jira_defect_metrics_updater` task. The information obtained is the timestamp when an issue change their status to `In Progress` and when the issue is finished (status `Done`).

Resolves [EM-226:](https://rootstrap.atlassian.net/browse/EM-226) (in part): See project metric: Development Cycle.

### CODE REVIEW

File: app/services/processors/jira_project_defect_escape_rate_updater.rb [hunk 1]
```diff
@@ -28,5 +25,18 @@ def call
...
+
+    def issue_update!(issue, bug_fields)
+      environment_field = bug_fields[ENV['JIRA_ENVIRONMENT_FIELD'].to_sym]
+      status_field = bug_fields[:status.to_sym][:name]
+
+      issue.update!(
+        informed_at: bug_fields[:created],
+        resolved_at: bug_fields[:resolutiondate] || nil,
+        in_progress_at: status_field == IN_PROGRESS ? bug_fields[:updated] : issue.in_progress_at,
+        environment: environment_field ? environment_field.first[:value].downcase : nil,
```
Comment (sebastiancaraballo) [inline app/services/processors/jira_project_defect_escape_rate_updater.rb:37]: I'd extract this logic into a method.

File: app/services/processors/jira_project_development_cycle_updater.rb [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+    end
+
+    private
+
+    def issues_to_update
+      JiraClient::Repository.new(@jira_project).issues
+    end
+
+    def issue_update!(issue, issue_fields)
+      environment_field = issue_fields[ENV['JIRA_ENVIRONMENT_FIELD'].to_sym]
```
Comment (sebastiancaraballo) [inline app/services/processors/jira_project_development_cycle_updater.rb:30]: same here

File: spec/services/processors/jira_project_defect_escape_rate_updater_spec.rb [hunk 3]
```diff
@@ -26,9 +30,10 @@
       it 'creates a record on the db' do
         expect { subject }.to change { JiraIssue.count }.from(0).to(1)
 
-        expect(JiraIssue.last.jira_project).to eq(jira_project)
-        expect(JiraIssue.last.informed_at).to eq('2021-03-14T15:48:04.000-0300')
-        expect(JiraIssue.last.environment).to eq('production')
+        expect(last_issue.jira_project).to eq(jira_project)
+        expect(last_issue.informed_at).to eq('2021-03-14T15:48:04.000-0300')
+        expect(last_issue.environment).to eq('production')
+        expect(last_issue.issue_type).to eq('bug')
```
Comment (sebastiancaraballo) [inline spec/services/processors/jira_project_defect_escape_rate_updater_spec.rb:36]: https://www.betterspecs.org/#single

File: app/services/processors/jira_project_development_cycle_updater.rb (hunk 2 above)
Comment (enzofab91) [reply app/services/processors/jira_project_development_cycle_updater.rb:30]: Done!

File: app/services/processors/jira_project_defect_escape_rate_updater.rb [hunk 4]
```diff
@@ -1,5 +1,7 @@
 module Processors
   class JiraProjectDefectEscapeRateUpdater < BaseService
+    IN_PROGRESS = 'In Progress'.freeze
```
Comment (lauperalti) [inline app/services/processors/jira_project_defect_escape_rate_updater.rb:3]: WDYT about putting this in a file with all the constants?

File: app/services/processors/jira_project_development_cycle_updater.rb [hunk 5]
```diff
@@ -0,0 +1,46 @@
+module Processors
+  class JiraProjectDevelopmentCycleUpdater < BaseService
+    IN_PROGRESS = 'In Progress'.freeze
```
Comment (lauperalti) [inline app/services/processors/jira_project_development_cycle_updater.rb:3]: same [here](https://github.com/rootstrap/rs-code-review-metrics/pull/449/files#r678568620)

**Comment by lauperalti** [comment]
Sorry for this:
Screen Shot 2021-07-28 at 19 16 29
I made a mistake!

BTW For me it is Approved, just take a look at why is falling the CI test.

**Review by lauperalti — APPROVED** [review]

File: app/services/jira_client/repository.rb [hunk 6]
```diff
@@ -5,17 +5,34 @@ def initialize(jira_project)
...
     rescue Faraday::ForbiddenError => exception
       ExceptionHunter.track(exception)
     end
+
+    private
+
+    def create_request(issue_type)
+      @request_params = {
+        jql: "project=#{@jira_project.jira_project_key} AND #{issue_type}",
+        fields: "#{ENV['JIRA_ENVIRONMENT_FIELD']},created"
```
Comment (sebastiancaraballo) [inline app/services/jira_client/repository.rb:28]: same here

File: app/services/processors/jira_project_development_cycle_updater.rb [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+        informed_at: issue_fields[:created],
+        resolved_at: issue_fields[:resolutiondate] || nil,
+        in_progress_at: status_field == IN_PROGRESS ? issue_fields[:updated] : issue.in_progress_at,
+        environment: environment_field ? environment_field.first[:value]&.downcase : nil,
+        issue_type: issue_type_field(issue_fields)&.downcase
+      )
+    end
+
+    def issue_type_field(issue_fields)
+      issue_fields[:issuetype.to_sym][:name]
```
Comment (sebastiancaraballo) [inline app/services/processors/jira_project_development_cycle_updater.rb:44]: 👀

**Review by sebastiancaraballo — APPROVED** [review]

File: app/services/processors/jira_project_development_cycle_updater.rb (hunk 7 above)
Comment (enzofab91) [reply app/services/processors/jira_project_development_cycle_updater.rb:44]: My bad, fix it 😄

**Pull request merged**